src/main/java/build/buildfarm/instance/shard/CasWorkerMap.java

```diff
@@ -0,0 +1,529 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.instance.shard;
+
+import build.bazel.remote.execution.v2.Digest;
+import build.buildfarm.common.DigestUtil;
+import build.buildfarm.common.redis.RedisClient;
+import com.google.common.collect.ImmutableMap;
+import java.io.IOException;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.redisson.api.RSetMultimapCache;
+import org.redisson.api.RedissonClient;
+import redis.clients.jedis.JedisClusterPipeline;
+
+///
+/// @class   CasWorkerMap
+/// @brief   A mapping from blob digest to the workers where the blobs
+///          reside.
+/// @details This is used to identify the location of blobs within the shard.
+///          {blob digest -> set(worker1,worker2)}.
+///
+public class CasWorkerMap {
```

Comment:
Looks like we are combining two separate implementations here. This would be a good candidate to separate these implementation into their own classes and utilize a correct one based on the useMemoryCache configuration.

Reply:
fixed